### PR TITLE
[FEATURE] add suggestions to phpcs Task (use phpcbf)

### PR DESCRIPTION
**Pull request by Kanti**
| Q             | A
| ------------- | ---
| Branch        | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Documented?   | no
| Fixed tickets | -

This adds phpcbf suggestions for the phpcs task.
it looks somthing like this:
````bash
FILE: /var/www/Classes/Command/CacheCommandController.php
----------------------------------------------------------------------
FOUND 4 ERRORS AFFECTING 3 LINES
----------------------------------------------------------------------
 28 | ERROR | [x] Opening brace of a class must be on the line after
    |       |     the definition
 36 | ERROR | [x] Opening brace should be on a new line
 49 | ERROR | [x] Expected 1 newline at end of file; 0 found
 49 | ERROR | [x] The closing brace for the class must go on the next
    |       |     line after the body
----------------------------------------------------------------------
PHPCBF CAN FIX THE 4 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: /var/www/Classes/Controller/BannerController.php
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 40 | ERROR | [x] TRUE, FALSE and NULL must be lowercase; expected
    |       |     "null" but found "NULL"
 76 | ERROR | [x] Expected 1 newline at end of file; 0 found
----------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------

Time: 3.33 secs; Memory: 38Mb


You can fix some errors by running following command:
'/var/www/vendor/bin/phpcbf' '--standard=PSR2' '--warning-severity=0' '--tab-width=4' '/var/www/Classes/Command/CacheCommandController.php' '/var/www/Classes/Controller/BannerController.php'
````

### CODE REVIEW

**Comment by veewee** [comment]
Thanks for the PR. It looks like there is an issue on Windows: https://ci.appveyor.com/project/veewee/grumphp/build/1.0.174
Can you take a look at it? 

Windows is using double quotes instead of single quotes for wrapping it's arguments.

**Comment by Kanti** [comment]
I will look at it.

**Comment by Kanti** [comment]
I have fixed it. When will you (merge) release the 0.9.2?

**Comment by veewee** [comment]
Currently I am on vacation. I'll take a deeper look in the code when I am on my computer. The new  release will be somewhere in the beginning of September.

File: src/GrumPHP/Formatter/PhpcsFormatter.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        $output = $process->getOutput();
+        if (!$output) {
+            return $process->getErrorOutput();
+        }
+
+        $pos = strrpos($output, "\n");
+        if ($pos === false) {
+            return $output;
+        } else {
```
Comment (veewee) [inline src/GrumPHP/Formatter/PhpcsFormatter.php]: The `else` statement is not necesarry

File: src/GrumPHP/Task/Phpcs.php [hunk 2]
```diff
@@ -70,21 +77,40 @@ public function run(ContextInterface $context)
...
+        $arguments->add('--report-json');
         $arguments->addFiles($files);
 
         $process = $this->processBuilder->buildProcess($arguments);
         $process->run();
 
         if (!$process->isSuccessful()) {
-            return TaskResult::createFailed($this, $context, $this->formatter->format($process));
+            $output = $this->formatter->format($process);
+            if (!$output) {
```
Comment (veewee) [inline src/GrumPHP/Task/Phpcs.php]: This part looks a bit strange. Why couldn't the formatter just format the message?

File: src/GrumPHP/Formatter/PhpcsFormatter.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    /**
+     * @param ProcessArgumentsCollection $defaultArguments
+     * @param ProcessBuilder $processBuilder
+     * @return string
+     */
+    public function formatErrorMessage(ProcessArgumentsCollection $defaultArguments, ProcessBuilder $processBuilder)
+    {
+        $defaultArguments->addArgumentArray('%s', $this->suggestedFiles);
```
Comment (veewee) [inline src/GrumPHP/Formatter/PhpcsFormatter.php]: This part is assuming that the `suggestedFiles` property is set. Isn't it better to run the `getSuggestedFilesFromJson()` method here?

**Comment by veewee** [comment]
It looks pretty good. I've added some small remarks.

File: src/GrumPHP/Formatter/PhpcsFormatter.php (hunk 3 above)
Comment (Kanti) [inline src/GrumPHP/Formatter/PhpcsFormatter.php]: At this point we cannot call ``getSuggestedFilesFromJson()`` because we don't have the Json.

**Comment by Kanti** [comment]
I have changed some stuff and hope it is now clearer.

File: src/GrumPHP/Formatter/PhpcsFormatter.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace GrumPHP\Formatter;
+
+use GrumPHP\Collection\ProcessArgumentsCollection;
+use GrumPHP\Process\ProcessBuilder;
+use Symfony\Component\Process\Process;
+use TYPO3\CMS\Extbase\Utility\DebuggerUtility;
```
Comment (veewee) [inline src/GrumPHP/Formatter/PhpcsFormatter.php]: This dependency is not part of GrumPHP and is not used. Can it be removed?

**Comment by veewee** [comment]
It's clear what the formatter is doing, but I think the spec for the formatter is not advanced enough:

- In the task you retrieve both the full report and the json report. This should also be very well tested in the specs.
- You also check if the json can be parsed. The result is empty, but you never tested if the file suggested files are valid.
- There is also no check if the phpcbf command is what you expect ut to be.

Can you take a look at this?

**Comment by Kanti** [comment]
I have added some Checks.

But I don't understand how we would check,
if the phpcbf command is what we expect it to be.

What exactly do you mean by this?

**Comment by veewee** [comment]
I meant that there was no test that checks that this returned what you expected:

```
$this->formatter->formatErrorMessage($arguments, $this->processBuilder);
```

Can you make the spec more understandable by using real data instead of something?
It's not clear how it should behave by just adding 'something'.

Sorry for the nitpicks, but I want to make sure that this code works exactly as it should :)

**Comment by Kanti** [comment]
> Sorry for the nitpicks, but I want to make sure that this code works exactly as it should :)

This is totally ok

File: spec/GrumPHP/Formatter/PhpcsFormatterSpec.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        ));
+
+        $arguments = new ProcessArgumentsCollection(array('phpcbf'));
+        $process->getOutput()->willReturn($this->getExampleData() . $json);
+        $this->format($process)->shouldBe($this->getExampleData());
+
+        $this->getSuggestedFilesFromJson(json_decode($json, true))->shouldBe(array('/var/www/Classes/Command/CacheCommandController.php'));
+
+        $processBuilder->buildProcess($arguments)->willReturn($process);
+        $process->getCommandLine()->willReturn();
```
Comment (veewee) [inline spec/GrumPHP/Formatter/PhpcsFormatterSpec.php]: Here you can return a sample commandline which you can check in the `shouldBe() ` of the `formatErrorMessage()` method  on the next line.

File: spec/GrumPHP/Formatter/PhpcsFormatterSpec.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        ));
+
+        $arguments = new ProcessArgumentsCollection();
+        $process->getOutput()->willReturn($this->getExampleData() . $json);
+        $this->format($process)->shouldBe($this->getExampleData());
+
+        $this->getSuggestedFilesFromJson(json_decode($json, true))->shouldBe(array('/var/www/Classes/Command/CacheCommandController.php'));
+
+        $processBuilder->buildProcess($arguments)->willReturn($process);
+        $process->getCommandLine()->shouldBeCalled();
```
Comment (veewee) [inline spec/GrumPHP/Formatter/PhpcsFormatterSpec.php]: See comment at line 75.

**Comment by veewee** [comment]
@Kanti: almost there, it's starting to look great! :)
I've added 2 comments in the spec file in which we can see what is the expected result of the suggestion part of the formatter. (Now it is just empty)

There is also still the comment on the unused Typo3 `DebuggerUtility` which is `use`'d in the task but never used.

**Comment by veewee** [comment]
Thanks for you work, it's a nice addition!

**Pull request merged**